File: old-projects/vector-defence-silverlight-port/src/Towers/GunTower.ts
import { Point } from "../Point";
import { Game } from "../Game";
import { Util } from "../Util";
import { Projectile } from "../Particles/Projectile";
import { Monster } from "../Monsters/Monster";
import { Tower } from "./Tower";

export class GunTower extends Tower {
  private static readonly gunLength = 16;
  private trackedMonster: Monster | null = null;
  private gunAngle = 0;
  private gunThickness = 2;

  constructor(x: number, y: number, addToGame = true) {
    super(x, y, addToGame);
    this.cost = 20;
    this.range = 60;
    this.millisecondsBetweenFiring = 200;
    this.gunAngle = Util.randomInRange(-Math.PI, Math.PI);
  }

  override render(context: CanvasRenderingContext2D): void {
    const end = this.gunEnd();
    this.drawBaseCircle(context, "#fff");
    context.save();
    context.strokeStyle = "#fff";
    context.lineWidth = this.gunThickness;
    context.lineCap = "round";
    context.beginPath();
    context.moveTo(this.x, this.y);
    context.lineTo(end.x, end.y);
    context.stroke();
    context.restore();
  }

  protected override animate(): void {
    if (this.trackedMonster === null || this.trackedMonster.markedForRemoval) {
      this.trackedMonster = this.getClosestMonsterInRange();
    }

    if (this.trackedMonster !== null) {
      const closeEnough = Util.isWithinDistance(this.trackedMonster.x, this.trackedMonster.y, this.x, this.y, this.range);
      if (!closeEnough || this.trackedMonster.markedForRemoval) {
        this.trackedMonster = null;
      } else {
        const source = this.gunEnd();
        const target = Tower.calculateWhereToShoot(source, this.trackedMonster, Projectile.speed);
        this.gunAngle = Util.calculateAngle(this.x, this.y, target.x, target.y);

        if (this.canFireNow) {
          const damageCaused = 10 + this.level;
          const size = 3 + (this.level / 2);
          new Projectile(source, target, damageCaused, size);
          this.timeFiring();
        }
      }
    }
  }

  protected override upgradeWork(newLevel: number): void {
    super.upgradeWork(newLevel);
    this.gunThickness = 2 + (newLevel / 2);
  }

  private gunEnd(): Point {
    return new Point(
      this.x + (GunTower.gunLength * Math.cos(this.gunAngle)),
      this.y + (GunTower.gunLength * Math.sin(this.gunAngle)),
    );
  }
}

File: old-projects/vector-defence-silverlight-port/original-csharp-source/Towers/GunTower.cs
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Shapes;
using VectorDefenceSL.Monsters;
using VectorDefenceSL.Particles;

namespace VectorDefenceSL.Towers
{
    public class GunTower : Tower
    {
        private const double GunLength = 16;

        private readonly Line gunLine = new Line
        {
            Stroke = new SolidColorBrush(Colors.White),
            StrokeThickness = 2,
            StrokeStartLineCap = PenLineCap.Round,
            StrokeEndLineCap = PenLineCap.Flat,
        };

        private readonly Ellipse torrentEllipse = new Ellipse
        {
            StrokeThickness = 1.5,
            Fill = new RadialGradientBrush(Colors.White, Colors.Black),
            Stroke = new SolidColorBrush(Colors.White),
            Width = TowerRadius * 2,
            Height = TowerRadius * 2
        };

        private Monster trackedMonster;
        private double gunAngle;

        public GunTower(Button btn) : base(btn, TowerBuildingMode.GunTower)
        {
        }

        public GunTower(double x, double y) : base(x, y)
        {
            Cost = 20;
            Range = 60;
            MillisecondsBetweenFiring = 200;
            torrentEllipse.MouseLeftButtonDown += HandleTowerSelection;
            gunLine.MouseLeftButtonDown += HandleTowerSelection;
        }

        public override void RemoveTowerFromCanvas()
        {
            base.RemoveTowerFromCanvas();
            Container.Children.Remove(torrentEllipse);
            Container.Children.Remove(gunLine);
        }

        public override sealed void Render()
        {
            gunLine.X1 = X;
            gunLine.Y1 = Y;
            gunAngle = Util.RandomInRange(-Math.PI, Math.PI);
            TurnGun();

            torrentEllipse.SetCanvasPosition(X - TowerRadius, Y - TowerRadius);
            Container.Children.Add(torrentEllipse);
            Container.Children.Add(gunLine);
        }

        protected override void Animate()
        {
            if (trackedMonster == null || trackedMonster.MarkedForRemoval)
            {
                trackedMonster = GetClosestMonsterInRange();
            }

            if (trackedMonster != null)
            {
                bool closeEnough = Util.IsWithinDistance(trackedMonster.X, trackedMonster.Y, X, Y, Range);
                if (!closeEnough || trackedMonster.MarkedForRemoval)
                {
                    trackedMonster = null;
                }
                else
                {
                    var source = new Point(gunLine.X2, gunLine.Y2);
                    var target = CalculateWhereToShoot(source, trackedMonster, Projectile.Speed);
                    gunAngle = Util.CalculateAngle(X, Y, target.X, target.Y);
                    TurnGun();

                    if (CanFireNow)
                    {
                        double damageCaused = 10d + Level;
                        double size = 3d + (Level / 2d);
                        new Projectile(source, target, damageCaused, size);
                        TimeFiring();
                    }
                }
            }
        }

        protected override void AnimateToolboxRendering()
        {
            gunAngle += ToolbarAnimationRotationStep * Game.CurrentFrameMultiplier;
            TurnGun();
        }

        protected override void UpgradeWork(int newLevel)
        {
            base.UpgradeWork(newLevel);
            gunLine.StrokeThickness = 2 + (newLevel / 2d);
            gunLine.Visibility = Visibility.Collapsed;
            gunLine.Visibility = Visibility.Visible;
        }

        private void TurnGun()
        {
            var dx = GunLength * Math.Cos(gunAngle);
            var dy = GunLength * Math.Sin(gunAngle);
            gunLine.X2 = X + dx;
            gunLine.Y2 = Y + dy;
        }
    }
}
